### PR TITLE
podman-search: run in parallel

### PR DESCRIPTION
**Pull request by vrothberg**
* Run image searches in parallel.
* Refactor code to be used by podman and the varlink API.
* Extend varlink API with an image search filter.

Fixes: #2345

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM

**Comment by vrothberg** [comment]
@giuseppe @baude @mheon PTAL

**Comment by mheon** [comment]
Does this actually fix the original issue? It sounded like they wanted
results streamed, as opposed to Varlink waiting for the entire process to
finish before returning.

Regardless, this is a good change - just maybe not the exact one they asked
for.

**Comment by vrothberg** [comment]
> Does this actually fix the original issue? It sounded like they wanted results streamed, as opposed to Varlink waiting for the entire process to finish before returning. Regardless, this is a good change - just maybe not the exact one they asked for.

That's a valid point. To preserve the order (and to avoid a random sequence) the "printing as soon as finished" would need to happen according the order of the registries. I am not sure that's worth the effort excluding cases where a few queries take unusually long.

**Comment by vrothberg** [comment]
Oh, wait. I missed the Varlink API point entirely. Thanks, @mheon!

**Comment by vrothberg** [comment]
To make this right, I need to refactor most parts into `libpod/image`. I am on it :)

File: cmd/podman/search.go [hunk 1]
```diff
@@ -172,72 +174,102 @@ func getSearchOutput(term string, registries []string, opts searchOpts, filter s
...
-	if err != nil {
-		return err
+	// searchOutputData is used as a return value for searching in parallel.
+	type searchOutputData struct {
+		data []searchParams
+		err  error
+	}
+
+	// Let's follow Firefox by limiting parallel downloads to 6.
+	sem := semaphore.NewWeighted(6)
```
Comment (TomSweeneyRedHat) [inline cmd/podman/search.go]: might be good to have '6' define/configurable somewhere in libpod.conf.

**Comment by vrothberg** [comment]
@mheon could you have another look. The code is now refactored, the Varlink API makes use of it and is extended with a filter. @jwhonce might be interested in this PR as well :)

File: libpod/image/search.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+	// Official indicates if it's an official image.
+	Official string
+	// Automated indicates if the image was created by an automated build.
+	Automated string
+}
+
+// SearchOptions are used to control the behaviour of SearchImages.
+type SearchOptions struct {
+	// Filter allows to filter the results.
+	Filter []string
```
Comment (mheon) [inline libpod/image/search.go]: Can we nuke this in favor of just making searchFilterParams public and embedding it?

File: libpod/image/search.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+	// Limit limits the number of queries per index (default: 25).
+	Limit int
```
Comment (mheon) [inline libpod/image/search.go]: What does 0 do here, no limit? Should we make this a *int so unset is the default of 25?

Comment (vrothberg) [reply libpod/image/search.go]: It must greater than 0 to overwrite the default. I prefer to keep it an `int` (or maybe `uint`?). I think it's worth extending the comment.

File: libpod/image/search.go (hunk 2 above)
Comment (vrothberg) [reply libpod/image/search.go]: What I like about the current version is that it avoids boilerplate code at the call sites, mainly to transform the `[]string` into the `searchFilterParams`. But if you prefer the other way, I can change that. Just let me know what you think after this comment.

Comment (mheon) [reply libpod/image/search.go]: Can we keep the parsing code common (maybe stick it in cmd/podman/shared?) so we can use the nicer searchFilterParams interface in libpod/image? It seems like a much nicer API for folks trying to use search

File: libpod/image/search.go (hunk 3 above)
Comment (mheon) [reply libpod/image/search.go]: Ack, if 0 is the default, that ought to be fine - just need to reflect it in the comment

File: libpod/image/search.go (hunk 2 above)
Comment (vrothberg) [reply libpod/image/search.go]: I think it's worth applying a similar change to the Varlink API changing the `[]string` to a `SearchFilter` (including a change from `*bool` to an opt bool).

Comment (vrothberg) [reply libpod/image/search.go]: Done

File: cmd/podman/search.go (hunk 1 above)
Comment (vrothberg) [reply cmd/podman/search.go]: We have the "magic" number when copying images in parallel, which makes it hard to control via libpod.conf (required some extension to the context).

However, I don't think that is something a user should worry about. 6 seems somehow random but seems to be proven value for nice throughput while playing nice with the servers.

**Comment by vrothberg** [comment]
@mheon @baude @TomSweeneyRedHat @rhatdan PTAL

**Comment by mheon** [comment]
Changes LGTM but would like a nod from @baude on the Varlink stuff

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM, but it looks like you're fighting some test unhappiness.

```
[Fail] Podman top [It] podman pod top on pod with containers in different namespace 
/var/tmp/go/src/github.com/containers/libpod/test/e2e/pod_top_test.go:133

[Fail] Podman checkpoint [It] podman checkpoint container with established tcp connections 
/var/tmp/go/src/github.com/containers/libpod/test/e2e/checkpoint_test.go:207
```

**Comment by vrothberg** [comment]
Tests are green now. Looked like a hiccup in the CI.

Now we need @baude's Varlink expertise :)

**Comment by rhatdan** [comment]
We really need the cockpit people to test this, but it LGTM.

**Comment by vrothberg** [comment]
> We really need the cockpit people to test this, but it LGTM.

We've tested via the Varlink CL. Currently, the PR does everything but to stream the results when they're available. I don't think that's necessary anymore due to the parallel execution. It also avoids complex code.

**Pull request merged**